File: Mailings/models/order_event.go
package models

import (
	"time"
)

type OrderEvent struct {
	EventId    string    `json:"eventId"`
	OccurredOn time.Time `json:"occurredOn"`
	OrderId    OrderId   `json:"orderId"`
}

type OrderId struct {
	Value int64 `json:"value"`
}

type CustomerId struct {
	Value int64 `json:"value"`
}

type Money struct {
	Amount   float64 `json:"amount"`
	Currency string  `json:"currency"`
}

type OrderCreatedEvent struct {
	OrderEvent
	CustomerId  CustomerId `json:"customerId"`
	TotalAmount Money      `json:"totalAmount"`
}

type OrderItemAddedEvent struct {
	OrderEvent
	Item OrderItem `json:"item"`
}

type OrderItemRemovedEvent struct {
	OrderEvent
	ItemId OrderItemId `json:"itemId"`
}

type OrderItemId struct {
	Value int64 `json:"value"`
}

type OrderItem struct {
	Id          OrderItemId `json:"id"`
	OrderId     OrderId     `json:"orderId"`
	ProductId   string      `json:"productId"`
	ProductName string      `json:"productName"`
	Quantity    int         `json:"quantity"`
	Price       Money       `json:"price"`
}

type OrderStatusChangedEvent struct {
	OrderEvent
	PreviousStatus string `json:"previousStatus"`
	NewStatus      string `json:"newStatus"`
}

type OrderCancelledEvent struct {
	OrderEvent
	PreviousStatus string `json:"previousStatus"`
	Reason         string `json:"reason"`
}

type OrderDiscountAppliedEvent struct {
	OrderEvent
	DiscountId         string  `json:"discountId"`
	DiscountName       string  `json:"discountName"`
	DiscountPercentage float64 `json:"discountPercentage"`
	OriginalAmount     Money   `json:"originalAmount"`
	DiscountedAmount   Money   `json:"discountedAmount"`
}

type OrderDiscountRemovedEvent struct {
	OrderEvent
	DiscountId       string `json:"discountId"`
	OriginalAmount   Money  `json:"originalAmount"`
	DiscountedAmount Money  `json:"discountedAmount"`
}

File: ProcessingOrders/ProcessingOrders.CoreDomain/Events/OrderEvent.cs
using System;
using ProcessingOrders.CoreDomain.Common;
using ProcessingOrders.CoreDomain.Entities;
using ProcessingOrders.CoreDomain.ValueObjects;

namespace ProcessingOrders.CoreDomain.Events;

public abstract class OrderEvent : IDomainEvent
{
    public Guid EventId { get; }
    public DateTime OccurredOn { get; }
    public long OrderId { get; }

    protected OrderEvent(long orderId)
    {
        EventId = Guid.NewGuid();
        OccurredOn = DateTime.UtcNow;
        OrderId = orderId;
    }
}

public class OrderCreatedEvent : OrderEvent
{
    public CustomerId CustomerId { get; }
    public Money TotalAmount { get; }

    public OrderCreatedEvent(long orderId, CustomerId customerId, Money totalAmount)
        : base(orderId)
    {
        CustomerId = customerId;
        TotalAmount = totalAmount;
    }
}

public class OrderStatusChangedEvent : OrderEvent
{
    public OrderStatus PreviousStatus { get; }
    public OrderStatus NewStatus { get; }

    public OrderStatusChangedEvent(long orderId, OrderStatus previousStatus, OrderStatus newStatus)
        : base(orderId)
    {
        PreviousStatus = previousStatus;
        NewStatus = newStatus;
    }
}

public class OrderItemAddedEvent : OrderEvent
{
    public OrderItem Item { get; }

    public OrderItemAddedEvent(long orderId, OrderItem item)
        : base(orderId)
    {
        Item = item;
    }
}

public class OrderItemRemovedEvent : OrderEvent
{
    public long ItemId { get; }

    public OrderItemRemovedEvent(long orderId, long itemId)
        : base(orderId)
    {
        ItemId = itemId;
    }
}

public class OrderCancelledEvent : OrderEvent
{
    public OrderStatus PreviousStatus { get; }
    public string Reason { get; }

    public OrderCancelledEvent(long orderId, OrderStatus previousStatus, string reason)
        : base(orderId)
    {
        PreviousStatus = previousStatus;
        Reason = reason;
    }
} 